pkg/filters/fields.go

```diff
@@ -81,15 +81,51 @@ type FieldFilter struct {
 	fields         fmutils.NestedMask
 	action         tetragon.FieldFilterAction
 	invertEventSet bool
+	needsCopy      map[tetragon.EventType]struct{}
 }
 
 // NewFieldFilter constructs a new FieldFilter from a set of fields.
 func NewFieldFilter(eventSet []tetragon.EventType, fields []string, action tetragon.FieldFilterAction, invertEventSet bool) *FieldFilter {
+	var maybeNeedsCopy bool
+	if action == tetragon.FieldFilterAction_EXCLUDE {
+		for _, field := range fields {
+			if field == "process.pid" || field == "parent.pid" {
+				maybeNeedsCopy = true
+				break
+			}
+		}
+	} else if action == tetragon.FieldFilterAction_INCLUDE {
+		var hasPid, hasPpid bool
+		maybeNeedsCopy = true
+		for _, field := range fields {
+			if field == "process.pid" {
+				hasPid = true
+			}
+			if field == "parent.pid" {
+				hasPpid = true
+			}
+		}
+		if hasPid && hasPpid {
+			maybeNeedsCopy = false
+		}
+	}
+
+	needsCopy := make(map[tetragon.EventType]struct{})
+	for _, t := range eventSet {
+		if t == tetragon.EventType_PROCESS_EXEC && maybeNeedsCopy {
```

Comment:
Why only check for exec and exit events? Wouldn't similar filters on all events have the same effect?

Reply:
Nope, the segfault is in the grpc/exec package which only deals with exec and exit events.

Reply:
Making a copy of every event will absolutely kill performance but yeah maybe it's the way to go. I think to be fully confident we'd need a full account of where exactly fields from events are used in the eventcache.

Reply:
The other option is refactoring so that this doesn't happen, but that is likely multiple weeks of work.

Reply:
> Making a copy of every event will absolutely kill performance but yeah maybe it's the way to go. I think to be fully confident we'd need a full account of where exactly fields from events are used in the eventcache.

I would argue that correctness is more important than performance.
 Also, this overhead would only be induced when  field filters are used, which I think is a reasonable tradeoff.